system/jlib/jmutex.hpp

```diff
@@ -541,6 +576,135 @@ class jlib_decl  NonReentrantSpinLock
+
+class CriticalBlockInstrumentation
+{
+#ifdef HAS_BACKTRACE
+    static constexpr unsigned numStackEntries = 8;
+    class Stack
+    {
+    public:
+        void *stack[numStackEntries];
+        bool operator==(const Stack &other) const
+        {
+            return memcmp(stack, other.stack, sizeof(stack)) == 0;
+        }
+    };
+
+    struct StackHash
+    {
+        std::size_t operator()(const Stack& k) const;
+    };
+#endif
+
+public:
+    CriticalBlockInstrumentation(const char *file, const char *func, unsigned line);
+    ~CriticalBlockInstrumentation();
+    void addStat(cycle_t wait, cycle_t hold, bool wasContended);
+    void describe(bool withStack) const;
+    int compare(const CriticalBlockInstrumentation *other) const;
+private:
+    NonReentrantSpinLock lock;
+    const char *file;
+    const char *func;
+    unsigned line;
+    cycle_t waitCycles;
+    cycle_t holdCycles;
+    unsigned uncontended;
+    unsigned contended;
+#ifdef HAS_BACKTRACE
+    std::unordered_map<Stack, unsigned, StackHash> stacks;
+#endif
+};
+
+class InstrumentedCriticalBlock
+{
+    InstrumentedCriticalSection &crit;
+public:
+    InstrumentedCriticalBlock(InstrumentedCriticalSection &c);
+    ~InstrumentedCriticalBlock();
+private:
+    CriticalBlockInstrumentation *inst;
+    cycle_t start = 0;
+    cycle_t in = 0;
+};
+
+/**
+ * Critical section delimiter, using scope to define lifetime of
+ * the lock on a critical section (parameter).
+ * Unblocks on construction, blocks on destruction.
+ */
+template<typename T> class CriticalUnblockOf
+{
+    T &crit;
+public:
+    inline CriticalUnblockOf(T &c) : crit(c) { crit.leave(); }
+    inline ~CriticalUnblockOf()              { crit.enter(); }
+};
+
+typedef CriticalUnblockOf<UninstrumentedCriticalSection> UninstrumentedCriticalUnblock;
+typedef CriticalUnblockOf<InstrumentedCriticalSection> InstrumentedCriticalUnblock;
+
+template<typename T> class CLeavableCriticalBlockOf
+{
+    T &crit;
+    bool locked = false;
+public:
+    inline CLeavableCriticalBlockOf(T &_crit) : crit(_crit)
+    {
+        enter();
+    }
+    inline CLeavableCriticalBlockOf(T &_crit, bool lock) : crit(_crit)
+    {
+        if (lock)
+            enter();
+    }
+    inline ~CLeavableCriticalBlockOf()
+    {
+        if (locked)
+            crit.leave();
+    }
+    inline void enter()
+    {
+        if (locked)
+            return;
+        locked = true;
+        crit.enter();
+    }
+    inline void leave()
+    {
+        if (locked)
+        {
+            locked = false;
+            crit.leave();
+        }
+    }
+};
+
+typedef CLeavableCriticalBlockOf<UninstrumentedCriticalSection> UninstrumentedLeavableCriticalBlock;
+typedef CLeavableCriticalBlockOf<InstrumentedCriticalSection> InstrumentedLeavableCriticalBlock;
+
+#ifdef USE_INSTRUMENTED_CRITSECS
+#define __glue(a,b) a ## b
+#define glue(a,b) __glue(a,b)
+
+extern thread_local CriticalBlockInstrumentation* __cbinst;
```

Comment:
Thread local variables generally generate terrible code.  It would be better to keep this hidden within jmutex.cpp and provide an accessor function.

Reply:
I think this is used to pass an implicit parameter into the CriticalBlock object without having to change the prototype?  If so it is worth adding a comment to explain.

Reply:
That is correct. I can add a comment.

Not really sure how an acessor function helps though - makes the code even more obscure, and less efficient (though smaller) I would have thought.